documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,823 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+1. The backend has a queued "mouse down" event.
+2. The backend updates the `State` of the relevant `IMouse` for that button press.
+3. The backend calls `HandleButtonDown` on the `IMouseInputHandler` (if applicable).
+4. The backend has a queued "mouse up" event.
+5. The backend updates the `State` of the relevant `IMouse` for that button release.
+6. The backend calls `HandleButtonUp` on the `IMouseInputHandler` (if applicable).
+
+This allows the actor to work with the whole device state with the device state being representative of the time that the original event occurred.
+
+More actors will be defined later in the proposal.
+
+All of the `Devices` and `Update`s are aggregated and coordinated by a central input context object.
+
+# Contexts
+
+```cs
+public partial class InputContext
+{
+    public Mice Mice { get; }
+    public Keyboards Keyboards { get; }
+    public Gamepads Gamepads { get; }
+    public Joysticks Joysticks { get; }
+    public IReadOnlyList<IInputDevice> Devices { get; }
+    public IList<IInputBackend> Backends { get; }
+    public event Action<ConnectionEvent>? ConnectionChanged;
+    public void Update();
+}
+```
+
+The central input object acts as the main entry point into the Input API, and is responsible for comparing the state reported by the devices for differences between `Update` calls (raising events as necessary). 
+
+`Mice`, `Keyboards`, `Gamepads`, and `Joysticks` are all custom `IReadOnlyList` types for enumerating the **state** (given that the actual input objects are thin, and the state contains a reference to the original device anyway). However, these custom types also contain the events. This is so we can "scope" the events, rather than putting them at the top-level and having to call the events `MouseButtonDown`, `JoystickButtonDown`, etc.
+
+By virtue of the `State` properties not updating until `IInputBackend.Update` is called, the states enumerated by the lists will not change until `Update` is called.
+
+`Update` will call `IInputBackend.Update` on each of the `Backends`, passing in a handler which implements `IInputHandler`, `IMouseInputHandler`, `IKeyboardInputHandler`, `IGamepadInputHandler`, and `IJoystickInputHandler` with each of the methods invoking a matching event defined in "Custom List Types".
+
+After the lists have been updated, the `Update` method compares the old state and the new state for that device and raises events accordingly. If there is no old state available for a device, the `ConnectionChanged` event (read on!) within each of the custom lists is called with a value of `true` signifying that a device has connected. Likewise, if there is no new state available for a device we have old state for, the `ConnectionChanged` event will be called with a value of `false` signifying that a device has been disconnected.
```

Comment:
> If there is no old state available for a device, the `ConnectionChanged` event...

Shouldn't we use the events from the handlers for this? Doesn't make much sense if the `InputContext` has to compare the old and new lists when the backend now tells us whenever a device connects or disconnects

Reply:
I agree, this is an outdated comment!
```suggestion
The central input object acts as the main entry point into the Input API, and is responsible for comparing the state reported by the devices for differences between `Update` calls (raising events as necessary). 

`Mice`, `Keyboards`, `Gamepads`, and `Joysticks` are all custom `IReadOnlyList` types for enumerating the **state** (given that the actual input objects are thin, and the state contains a reference to the original device anyway). However, these custom types also contain the events. This is so we can "scope" the events, rather than putting them at the top-level and having to call the events `MouseButtonDown`, `JoystickButtonDown`, etc.

By virtue of the `State` properties not updating until `IInputBackend.Update` is called, the states enumerated by the lists will not change until `Update` is called.

`Update` will call `IInputBackend.Update` on each of the `Backends`, passing in a handler which implements `IInputHandler`, `IMouseInputHandler`, `IKeyboardInputHandler`, `IGamepadInputHandler`, and `IJoystickInputHandler` with each of the methods invoking a matching event defined in "Custom List Types" or on the input context itself (such as `ConnectionChanged`).
```